src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -341,72 +343,97 @@ private void WatchForFileSystemEventsThreadStart(object arg)
-                            if (pairedId == long.MinValue)
-                            {
-                                // Getting here means we have a rename without a pair, meaning it should be a create for the 
-                                // move from unwatched folder to watcher folder scenario or a move from the watcher folder out.
-                                // Check if the item exists on disk to check which it is
-                                WatcherChangeTypes wct = DoesItemExist(eventPaths[i], IsFlagSet(eventFlags[i], Interop.EventStream.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile)) ?
-                                    WatcherChangeTypes.Created : 
-                                    WatcherChangeTypes.Deleted;
-                                watcher.NotifyFileSystemEventArgs(wct, relativePath);
-                            }
-                            else
-                            {
-                                // Remove the base directory prefix and add the paired event to the list of 
-                                // events to skip and notify the user of the rename 
-                                string newPathRelativeName = eventPaths[pairedId].Remove(0, _fullDirectory.Length);
-                                watcher.NotifyRenameEventArgs(WatcherChangeTypes.Renamed, newPathRelativeName, relativePath);
+                            watcher.NotifyFileSystemEventArgs(WatcherChangeTypes.Changed, relativePath);
+                        }
+
+                        // Disallow event coalescing. OSX likes to add a Created event to everything, particularly Changed events.
+                        if ((eventFlags[i] & changedFlags) > 0)
+                            return;
 
-                                // Create a new list, if necessary, and add the event
-                                if (handledRenameEvents == null)
+                        // Notify created/deleted/renamed events if they pass through the filters
+                        bool allowDirs = (_filterFlags & FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsDir) > 0;
+                        bool allowFiles = (_filterFlags & FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile) > 0;
+                        bool isDir = (eventFlags[i] & FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsDir) > 0;
+                        bool isFile = (eventFlags[i] & FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile) > 0;
+                        bool eventIsCorrectType = (isDir && allowDirs) || (isFile && allowFiles);
+                        bool eventIsLink = (eventFlags[i] & (FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsHardlink | FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsSymlink | FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsLastHardlink)) > 0;
+
+                        if (eventIsCorrectType || ((allowDirs || allowFiles) && (eventIsLink)))
```

Comment:
Yes. If you pass only a changed NotifyFilter to your watcher e.g. ```watcher.NotifyFilters = NotifyFilter.Attribute```  then these bools are always false and Created/Deleted/Renamed events get filtered out.